Add support for IKEv2 to ikebuster

CrsiX opened the pull request:
This PR adds IKEv2 support to the ISAKMP library and the `ikebuster` tool.

It may need fine-tuning here and there, but it works overall.

CrsiX committed:
Started building IKEv2 definitions

CrsiX committed:
Added remaining IKEv2 params as Rust enums

CrsiX committed:
Added NotifyMessageStatus

CrsiX committed:
Added basic generators, extended the definitions

CrsiX committed:
Added high-level representations and header structs

CrsiX committed:
Implemented `build` for `Transform`

CrsiX committed:
Implemented `Proposal.build` and some tests for it

CrsiX committed:
Implemented `SecurityAssociation.build`

CrsiX committed:
Implemented `IKEv2.build`, use key_length as u16 directly for encryption

CrsiX committed:
Implemented `Payload.build`, added KeyExchange type, simplified Nonce and VendorID

CrsiX committed:
Added NotifyHeader, Notification payload and its `build` method

CrsiX committed:
Extended IKEv2 packet generation

CrsiX committed:
Started building the parser package

CrsiX committed:
Switch from Vec<Transform> to 5 split up Vecs, fixed generation, fixed tests

CrsiX committed:
Built parser for SA

CrsiX committed:
Started parsing for a proposal

CrsiX committed:
Added proposal parser, fixed many things

CrsiX committed:
Fixed SA parser

CrsiX committed:
Switch from &u8 to Vec<u8> for notification data, implement Notify parser

CrsiX committed:
Introduce BoundaryErrors, made sure SA work with random attributes

CrsiX committed:
Moved tests to a separate file

CrsiX committed:
Fixed tests, added 'log', used try_build for Proposals

CrsiX committed:
Renamed params to use more CAPSLOCK_PASCAL_CASE for naming scheme

CrsiX committed:
Implement parsing for KeyExchange payload

CrsiX committed:
Added CertificateRequest parsing, fixed unknown payload bug of endless loop

CrsiX committed:
Moved constants into their own file, added IKEv2::hello

CrsiX committed:
Added small utils

CrsiX committed:
Disabled doctests on the library since they are broken due to formatting

CrsiX committed:
Added methods to get handshake or key lengths

CrsiX committed:
Added 'requires_integrity_algorithm', removed ENCR algos not allowed in IKE

CrsiX committed:
Start building ikebuster v2 support

CrsiX committed:
Mark encryption ciphers not allowed in IKE as such but allow their use

CrsiX committed:
Enforce integrity algorithms if no AEAD ciphers in a proposal

CrsiX committed:
Added 'list_all_proposals' to ikebuster for IKEv2

CrsiX committed:
Added setup description, added default proposal, added CSV formatter

CrsiX committed:
Added the Delete payload, header, generator and parser

CrsiX committed:
Added 'list_all_proposals' and a formatter to new v2_utils package

CrsiX committed:
Added first Scanner iteration

CrsiX committed:
Avoid zero-init-SPI bug

CrsiX committed:
RFC 7296, section 3.3.1 allows numbering start of Proposal other than 1 for the responder

CrsiX committed:
Implement the handler logic for the scanner

CrsiX committed:
Various smaller fixes

CrsiX committed:
Added finding support to scanner

CrsiX committed:
Added more elaborate detection for dropped packets & retries

CrsiX committed:
Added a serialization format and periodic saves to the scanner to avoid data loss

CrsiX committed:
Added a periodic state saving feature to the scanner

CrsiX committed:
Various extensions and added timeout support

CrsiX committed:
Added VendorID support

CrsiX committed:
Started a new scanner2 implementation, added CLI arguments to work with mode V2

CrsiX committed:
Separated receiver and sender

CrsiX committed:
Reworked the main function to call the tool successfully

CrsiX committed:
Added a busy loop to the main function for printing statistcs

CrsiX committed:
Improved the handling of the main v2

CrsiX committed:
Added instrument, improved timeout handling, implemented the receiver

CrsiX committed:
Handle verify packets, handle InvalidKeyExchange payloads

CrsiX committed:
Improve speed by accepting NO_PROPOSAL_CHOSEN for more than a single response

CrsiX committed:
Added printing findings to stdout

CrsiX committed:
Enable Path MTU Discovery by retrying failed packets

CrsiX committed:
Fixed printing, ensure at least one transform per message

CrsiX committed:
Discontinue 'v2' module

CrsiX committed:
Limit outgoing packets to the IKEv2 min msg size, retry on MTU changes

CrsiX committed:
Enable the new scanner, fixed a packet length bug

CrsiX committed:
Renamed v2_utils to v2 and moved the old v2 to v2_old

CrsiX committed:
Removed old v2 implementation for ikebuster

CrsiX committed:
Added more derive's, moved utility function for randomness into ikebuster

CrsiX committed:
Fixed DoS cookie handling because it is required to be the FIRST payload

CrsiX committed:
Renamed 'ModP' -> 'MODP' in accordance with the RFC

CrsiX committed:
Added half-open connections, moved more scan logic into the scanner module

CrsiX committed:
Implemented peeking, removed DELETE notifications, fixed bug that ignored half of the proposals

CrsiX committed:
Enabled peeking, changed output format for consistency

CrsiX committed:
Added a fallback for lost packets along the way but the host is alive

CrsiX committed:
Moved the main v1 implementation to its own function

CrsiX committed:
Implemented simple 'both' mode

CrsiX committed:
Moved the socket binding to its own utility

CrsiX committed:
Removed old scanner implementation

CrsiX committed:
Implemented the autodetection scan variant

CrsiX committed:
Added lots of docs, fixed imports, fixed timeout lengths

CrsiX committed:
Run cargo update

myOmikron commented on ikebuster/src/lib.rs:
```diff
@@ -12,7 +12,9 @@ use std::net::SocketAddr;
 use std::time::Duration;
 
-use isakmp::v1::definitions::NotifyMessageType;
+use isakmp::v1::definitions::{
+    AuthenticationMethod, EncryptionAlgorithm, GroupDescription, HashAlgorithm, NotifyMessageType,
+};
```
I've added a `.rustfmt.toml` for formatting guidelines, please use these

myOmikron commented on ikebuster/src/v2/finding.rs:
```diff
@@ -0,0 +1,99 @@
+                "".to_string()
+            },
+            f.kex,
+            f.result
```
Use named format variables

CrsiX committed:
Merge remote branch 'upstream/main'

CrsiX committed:
Run cargo update

CrsiX committed:
Reformatted imports of isakmp

CrsiX committed:
Reformatted imports of ikebuster

CrsiX committed:
Improve readability through format variables

CrsiX committed:
Added Unicode-3.0 license to allowlist

CrsiX committed:
Added missing SHA1 for peeking proposal

CrsiX committed:
Added support for Windows status codes from private use range

CrsiX committed:
Improve peeking with better info message

CrsiX committed:
Fixed a bug where open proposals would not be matched for overspecific key sizes

CrsiX committed:
Relax logging for informational requests from the responder

CrsiX committed:
Fixed duplicate findings, fixed local proposals that could be different

myOmikron commented on ikebuster/src/lib.rs:
```diff
@@ -231,4 +262,65 @@ pub enum ScanError {
+        Ok(Some(Ok(_)))
+    );
+    rx_handle.abort();
+    let _ = rx_handle.await;
```
Not necessary

myOmikron commented on ikebuster/src/main.rs:
```diff
@@ -73,22 +114,152 @@ pub struct DataOutput {
+                    if let Some(state) = handler.dump_state().await {
+                        let mut file = File::create(json_state_path)?;
+                        let content = serde_json::to_string_pretty(&state)?;
+                        let _ = file.write(content.as_bytes())?;
```
Use `to_writer_pretty()` from serde_json. `write` doesn't guarantee that all bytes have been written.

myOmikron commented on ikebuster/src/main.rs:
```diff
@@ -73,22 +114,152 @@ pub struct DataOutput {
+    if let Some(csv_path) = &cli.csv {
+        let mut file = File::create(csv_path)?;
+        let content = v2::finding::format_to_csv(&findings)?;
+        let _ = file.write(content.as_bytes())?;
```
use `write_all`

myOmikron commented on ikebuster/src/main.rs:
```diff
@@ -73,22 +114,152 @@ pub struct DataOutput {
+                .collect(),
+            vendor_ids: results.vendor_ids,
+        })?;
+        let _ = file.write(content.as_bytes())?;
```
Use `write_all`

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+        Transform::KeyExchange(KeyExchangeMethod::MODP_4096),
+        Transform::KeyExchange(KeyExchangeMethod::Curve_448),
+        Transform::KeyExchange(KeyExchangeMethod::Curve_25519),
+    ]);
```
Add the construction of your Proposals as method on the `Proposal`, so this method can be read without this huge chunks.

CrsiX replied:
I've split them up, which should make it more readable.

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+/// but if this returns `false` then the scan might not produce any results.
+/// Note that this should be done at the start of the scan, because it can quickly
+/// determine if a host is dead or does not support [IKEv2] at all.
+pub(crate) async fn peek(socket: &UdpSocket) -> Result<bool, ScanError> {
```
Don't name that peek - peek is used for working with iterators - I would suggest `probe_host`

CrsiX replied:
Using `probe_target` in a mod `probing`

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+            .map_err(ScanError::Send)?;
+        debug!(
+            "Sent {} bytes with huge proposal to check availability",
+            sent_bytes
```
Inline parameters where applicable

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+    ]);
+    let proposals = vec![p1, p2];
+
+    if let Some((msg, _)) = make_new_hello_packet(proposals) {
```
If this is always correct, use `expect`

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+
+        let mut accepted_indicators = 0;
+        let mut recv_buffer = [0u8; MAX_DATAGRAM_SIZE];
+        match socket.recv(&mut recv_buffer).await {
```
Early return in this match to reduce indentation levels

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+        let mut recv_buffer = [0u8; MAX_DATAGRAM_SIZE];
+        match socket.recv(&mut recv_buffer).await {
+            Ok(recv_bytes) => {
+                debug!(
+                    data = ?&recv_buffer[..recv_bytes],
+                    "Received {} bytes from responder",
+                    recv_bytes
+                );
```
Split into two logging statements:

debug: received X bytes
trace: received this bytes: <BYTES>

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+            sent_bytes
+        );
+
+        let mut accepted_indicators = 0;
```
This should be a boolean. 
Your neutral case can never occur and in case of an error or unexpected value, you set the value to -100, which can never be reached by the `+= 1` calls.

myOmikron commented on ikebuster/src/v2/peeking.rs:
```diff
@@ -0,0 +1,176 @@
+                        } else {
+                            error!(err = ?err, "Failed to parse IKEv2 packet: {}", err);
+                        }
+                        accepted_indicators -= 100;
```
Just early return in these cases

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                    results.vendor_ids.push(v);
+                }
+            }
+            _ => {}
```
Write all other cases explicitly, don't use `_`, so its easier to read and less likely to brake

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                        modified_proposal
+                            .encryption_algorithms
+                            .iter_mut()
+                            .for_each(|e| e.1 = None);
```
Just use a plain old for loop

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                            .iter_mut()
+                            .for_each(|e| e.1 = None);
+                        for mut p in open_proposals {
+                            p.encryption_algorithms.iter_mut().for_each(|e| e.1 = None);
```
Just use a plain old for loop

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                            .encryption_algorithms
+                            .iter_mut()
+                            .for_each(|e| e.1 = None);
+                        for mut p in open_proposals {
```
Don't use `mut p in xyz`, modern Rust likes to have: `p in &mut xyz`

CrsiX replied:
I consume `open_proposals`, and thus I'd say `mut p` is acceptable

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                        }
+                    }
+                    open.verify.push(a);
+                    open.verify.push(b);
```
Variable naming!

Using a placeholder variable  like `x` in the same line it is used, can be okay - multiple variables isn't okay

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+) {
+    if let Some(&sa) = open_packet
+        .payloads
+        .iter()
```
use `into_iter`, you have ownership

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+            _ => None,
+        })
+        .collect::<Vec<_>>()
+        .first()
```
Don't collect, just use `next`

myOmikron commented on ikebuster/src/v2/receiver.rs:
```diff
@@ -0,0 +1,306 @@
+                        } else {
+                            b.push(x.clone());
+                        }
+                    }
```
```
let [a, b] = v.split_off(v.len()/2);
```

is shorter

CrsiX replied:
Using `split_off` now.

CrsiX committed:
Relax the parsing to allow a little bit of RFC-noncompliance where it doesn't hurt much

CrsiX committed:
Modify acceptance indicators to differ between soft fail and hard fail

CrsiX committed:
Renamed peeking -> probing, fixed receiver splitting, integrate reviewer feedback

CrsiX committed:
Added a fix to detect invalid syntax in Microsoft notification results

CrsiX committed:
Increased the nonce size to 512 bits due to RFC 7296 section 2.10

It mandates that the nonce is at least the key size of the PRF,
which is at most 512 bits for SHA512.

CrsiX committed:
Renamed peeking to probing

CrsiX committed:
Fixed the ordering of the probing request to check non-AEAD algos first

CrsiX committed:
Changed ordering of proposals in probing, added AES_XCBC algorithm
